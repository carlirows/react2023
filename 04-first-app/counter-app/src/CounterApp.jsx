// functional component
import React from 'react';
import PropTypes from 'prop-types';
import { useState } from 'react';

export const CounterApp = ({value}) => {
    //add state to handle value
    const [counter, setCounter] = useState({
        counter1: value
    });
    
    const reset = () => {
        setCounter({
            ...counter,
            counter1: value
        });
    }

    const handleClick = (e) => {
        e.preventDefault();

        if(e.target.textContent === '-1'){
            setCounter({
                ...counter,
                counter1: counter.counter1 - 1
            });
            return;
        } 
        setCounter({
            ...counter,
            counter1: counter.counter1 + 1
        });

        return;
        
    }

    return (
        <>
            <h1>CounterApp</h1>
            <h2>{counter.counter1}</h2>
            <button onClick={handleClick}>-1</button>
            <button aria-label='reset' onClick={reset}>reset</button>
            <button onClick={handleClick}>+1</button>
        </>
    );
}



CounterApp.propTypes = {
    value: PropTypes.number
}
CounterApp.defaultProps = {
    value: 0
}