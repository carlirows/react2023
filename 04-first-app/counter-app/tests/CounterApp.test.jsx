import { fireEvent, render, screen } from "@testing-library/react";
import {CounterApp} from '../src/CounterApp.jsx';

describe('CounterApp tests', () => {

    test('should match with snapshot', () => {
        const {container} = render(<CounterApp value={0}/>);
        expect(container).toMatchSnapshot();
    });

    test('should display value sent by props', () => {
        const value = 100;
        render(<CounterApp value={value}/>);
        expect(screen.getByRole('heading', { level : 2 }).innerHTML).toBe(value.toString());
    });

    test('should increment with +1 button', () => {
        render(<CounterApp />);
        fireEvent.click(screen.getByText('+1'));
        expect(screen.getByRole('heading', { level : 2 }).innerHTML).toBe('1');
    });

    test('should decrement with -1 button', () => {
        render(<CounterApp />);
        fireEvent.click(screen.getByText('-1'));
        expect(screen.getByRole('heading', { level : 2 }).innerHTML).toBe('-1');
    });

    test('should reset with reset button', () => {
        render(<CounterApp />);
        fireEvent.click(screen.getByText('+1'));
        fireEvent.click(screen.getByText('+1'));
        fireEvent.click(screen.getByText('+1'));
        //fireEvent.click(screen.getByText('reset'));
        fireEvent.click(screen.getByRole('button', { name : 'reset' }));
        expect(screen.getByText(0)).toBeTruthy();
    });
});