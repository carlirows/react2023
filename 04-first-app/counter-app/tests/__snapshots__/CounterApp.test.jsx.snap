// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CounterApp tests should match with snapshot 1`] = `
<div>
  <h1>
    CounterApp
  </h1>
  <h2>
    0
  </h2>
  <button>
    -1
  </button>
  <button
    aria-label="reset"
  >
    reset
  </button>
  <button>
    +1
  </button>
</div>
`;
